### PR TITLE
Refine configuration design docs

### PR DESCRIPTION
**Pull request by pellared**
## Why

Get rid of:
- unusable exported API
- unnecessary noise in GoDoc

Make the docs a little less confusing.

Reference: https://github.com/open-telemetry/opentelemetry-go/issues/536#issuecomment-825731830

## What

Changes in [configuration deisn docs](https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#configuration).

### Unexport Apply method

I see no reason to export `Apply` the method in `Option` interfaces.
Unexporting the method is also an additional safety net if someone accidentally exports the `config` struct.

Example in practice: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/750

### Change T* to T

I find that using `T*` for a generic type name may be confusing. Someone may think it is `*T` when reading quickly (pointer to T). I suggest simply using `T` as a placeholder.

### CODE REVIEW

**Comment by pellared** [comment]
Can someone add `Skip Changelog` label, please? 😉

File: CONTRIBUTING.md
Comment (XSAM) [inline CONTRIBUTING.md:232]:
```suggestion
Moreover, the interface becomes sealed so the user cannot easily implement
```

**Review by XSAM — APPROVED** [review]

File: CONTRIBUTING.md
Comment (pellared) [reply CONTRIBUTING.md]: Fixed: https://github.com/open-telemetry/opentelemetry-go/pull/1841/commits/ed7502044ca5e05857525beb47311ed8f37a2336

**Review by MrAlias — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]

File: CONTRIBUTING.md
Comment (pellared) [reply CONTRIBUTING.md]: Thank you
Fixed: https://github.com/open-telemetry/opentelemetry-go/pull/1841/commits/eee94aadedf776dbc3510fccb1171c7e29e7f59b

File: CONTRIBUTING.md
Comment (pellared) [inline CONTRIBUTING.md:320]: In some places `\t` was used for indentation in others there were 4 spaces.

I decided to go with 4 spaces everywhere because my markdown linter does not complain when spaces are used 😉

Comment (MrAlias) [reply CONTRIBUTING.md:320]: Can we stick with hard tabs for the Go code? That way when people copy it it will be validly formatted code.

As for markdownlint we just merged #1842 which should configure your IDE to ignore those hard tabs when this is updated to the main branch.

Comment (pellared) [reply CONTRIBUTING.md:320]: Sure. I will fix it today.

Comment (pellared) [reply CONTRIBUTING.md]: Changed here and in other places as well: https://github.com/open-telemetry/opentelemetry-go/pull/1841/commits/1bc2d83cf2bd44ba8b9f1a4344aeb555f47f9f26

**Review by MrAlias — COMMENTED** [review]
The scope of the PR has crept beyond the initial scope of just addressing the content of the configuration style guide. I think the generalized markdown formatting changes are a great addition, but can we move them to another PR?

**Comment by pellared** [comment]
> a great addition, but can we move them to another PR?

Do you have a strong opinion here? 😉  I won't be able to fix it today and I think we would lose more time on the PRs than it is worth.

**Comment by MrAlias** [comment]
> > a great addition, but can we move them to another PR?
> 
> Do you have a strong opinion here?  I won't be able to fix it today and I think we would lose more time on the PRs than it is worth.

Without splitting things out this PR needs to have all the approvers re-evaluate their approval as the scope and content have changed.  I think it is probably the best approach to revert the changes here and add them to another PR. Once the revert is done here we can merge this.

**Comment by MrAlias** [comment]
We have a bit of an honor system around re-requesting reviews after substantive changes have been made to a PR: https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#how-to-get-prs-merged

If you want to keep the PR as is, please re-request reviews.

**Comment by pellared** [comment]
>  the best approach to revert the changes here and add them to another PR. Once the revert is done here we can merge this.

Done 😉 I will create a PR for reformatting all Go markdown code blocks once this is merged.

**Pull request merged**
